crates/sui-core/src/traffic_controller/mod.rs

```diff
@@ -112,74 +143,91 @@ impl TrafficController {
-                // TODO: once we've verified this doesn't happen under normal
-                // conditions, we can consider dropping the request itself given
-                // that clearly the system is overloaded
-            }
-            Err(TrySendError::Closed(_)) => {
-                panic!("TrafficController tally channel closed unexpectedly");
+        if let Some(channel) = self.tally_channel.as_ref() {
+            // Use try_send rather than send mainly to avoid creating backpressure
+            // on the caller if the channel is full, which may slow down the critical
+            // path. Dropping the tally on the floor should be ok, as in this case
+            // we are effectively sampling traffic, which we would need to do anyway
+            // if we are overloaded
+            match channel.try_send(tally) {
+                Err(TrySendError::Full(_)) => {
+                    warn!("TrafficController tally channel full, dropping tally");
+                    self.metrics.tally_channel_overflow.inc();
+                    // TODO: once we've verified this doesn't happen under normal
+                    // conditions, we can consider dropping the request itself given
+                    // that clearly the system is overloaded
+                }
+                Err(TrySendError::Closed(_)) => {
+                    panic!("TrafficController tally channel closed unexpectedly");
+                }
+                Ok(_) => {}
             }
-            Ok(_) => {}
         }
     }
 
     /// Handle check with dry-run mode considered
     pub async fn check(&self, client: &Option<IpAddr>, proxied_client: &Option<IpAddr>) -> bool {
-        match (
-            self.check_impl(client, proxied_client).await,
-            self.dry_run_mode(),
-        ) {
-            // check succeeded
-            (true, _) => true,
-            // check failed while in dry-run mode
-            (false, true) => {
-                debug!(
-                    "Dry run mode: Blocked request from client {:?}, proxied client: {:?}",
-                    client, proxied_client
-                );
-                self.metrics.num_dry_run_blocked_requests.inc();
-                true
+        let check_with_dry_run_maybe = |allowed| -> bool {
+            match (allowed, self.dry_run_mode()) {
+                // check succeeded
+                (true, _) => true,
+                // check failed while in dry-run mode
+                (false, true) => {
+                    debug!("Dry run mode: Blocked request from client {:?}", client);
+                    self.metrics.num_dry_run_blocked_requests.inc();
+                    true
+                }
+                // check failed
+                (false, false) => false,
+            }
+        };
+
+        match &self.acl {
+            Acl::Allowlist(allowlist) => {
+                let allowed = client.is_none() || allowlist.contains(&client.unwrap());
```

Comment:
will be `client` be `None`?

Reply:
Yeah the client should never be `None` in the happy path. This is just edge case handling for the case where the client parsing fails earlier in the flow. In such a case, we don't want to panic, since there have been some interesting cases in the past where this has failed in production after rollout, so we just set the client to `None` and essentially allow everything. There's logging and a metric that we bump (again, earlier in the flow) if that happens though so that we can investigate

Reply:
I wonder when it's `None` and it's in allowlist mode whether we should reject it as opposed to let it pass. Will leave it up to you

Reply:
I don't think there's a way that this can be made `None` by an attacker. It would generally mean that we have a bug in parsing, or traffic controller on the server itself is misconfigured. So it feels safer in this case to allow. If it's a bug, then we definitely don't want to be sinking all requests on the node, and if it's a misconfigured server, then that should be isolated to just that one node, so it should be safe from the perspective of the bridge network to just allow it. The worst case scenario if we reject in this case is that it's a bug that affects all nodes, and we effectively halt the bridge.